Signup using invite

yogirajh007 opened the pull request:
Solves issue #36

yogirajh007 committed:
Add invite-id to users spec

Some factory functions now do db queries as side effect

Using new factory functions in invites-tests

yogirajh007 committed:
Add reCAPTCHA check for user signup
Add db & model tests for user-invites
Add check for existance of org-id while creating new invite

yogirajh007 committed:
Fix tests to accomodate recaptcha
- When an exception is thrown recaptcha-invalid? returns true
- Change spec used in user-model/create
Fix middleware tests wrt reCAPTCHA
- Fix tests such that user without org is unable to create invites/

yogirajh007 committed:
Add RECAPTCHA secret key env variable

yogirajh007 committed:
Add more tests for user db & models wrt invites

yogirajh007 committed:
Remove bug in factories/invite-driver

yogirajh007 committed:
Add handler for signup-using-invite handler
- Add spec for users/org-id
- Add new spec - user-with-invite-token
- Modify user-db/create to handle new spec as well
- Modify views/signup-user to add hidden token input

alishamohanty commented on test/winter_onboarding_2021/fleet_management_service/handlers/invite_test.clj:
```diff
@@ -5,19 +5,24 @@
+                                             :usage-limit "10"}
+                               :headers {"host" "https://foober.com"}})
           db-resp (invite-db/find-by-keys {:invites/created-by (:users/id user)})]
       (is (= 302
```
what does this check?

alishamohanty commented on test/winter_onboarding_2021/fleet_management_service/handlers/invite_test.clj:
```diff
@@ -47,16 +52,18 @@
       (is (= 302
```
same, it should check with the status code of the response.

alishamohanty reviewed (approved):
Can u add screenshot of the view?
Can u change the commit message of `9ee032946e4db9190bece91a67bee04a4958e2d4` to rewrite the spelling of accommodate ?

yogirajh007 commented on src/winter_onboarding_2021/fleet_management_service/handlers/user.clj:
```diff
@@ -34,20 +36,45 @@
+         (response/redirect redirect-url)))
+
+(defn- create-user-via-invite [token validated-user]
+  (let [token-data (first (invite/find-by-keys {:invites/token token}))]
```
It contains not just the token, but other information as well, like role, created_by, etc

yogirajh007 commented on test/winter_onboarding_2021/fleet_management_service/views/user_test.clj:
```diff
@@ -13,8 +13,13 @@
-             (hf/hiccup-text (hf/hiccup-find [:#password] signup-form-output)))))))
-
+             (hf/hiccup-text (hf/hiccup-find [:#password] signup-form-output))))))
+  (testing "Should return extra input of token"
```
Any suggestions?

alishamohanty replied:
`should return hidden input field of token`
I am not sure if adds a lot of clarity but might make little more sense. What do u think?

yogirajh007 commented on src/winter_onboarding_2021/fleet_management_service/handlers/user.clj:
```diff
@@ -34,20 +36,45 @@
   (let [ns-form-params (utils/namespace-keys :users form-params)
         validated-user (s/conform ::specs/signup-form ns-form-params)
-        g-recaptcha-response (:g-recaptcha-response form-params)]
+        g-recaptcha-response (:g-recaptcha-response form-params)
```
g-recaptcha-response is consistent with Google docs.

yogirajh007 commented on test/winter_onboarding_2021/fleet_management_service/db/user_test.clj:
```diff
@@ -79,3 +79,13 @@
+        new-user (factories/user {:users/role (:invites/role invite)
+                                  :users/invite-id (:invites/id invite)})
+        resp (user-db/create new-user)
+        _ (println resp)]
```
Are you referring to an older commit?

yogirajh007 commented on src/winter_onboarding_2021/fleet_management_service/views/user.clj:
```diff
@@ -2,32 +2,35 @@
+            (when (some? token)
+              [:input {:type "hidden" :name "token" :value token :id "token"}])
+            [:div {:class "g-recaptcha"
+                   :data-sitekey "6Lc1XSkeAAAAAJUFg_Zv7qQ7w4bZ2v2sH7Y_cNW3"}]
```
This is not a secret, its public.

alishamohanty replied:
Even if it is not a secret, we should consider to add it in config instead.

yogirajh007 commented on test/winter_onboarding_2021/fleet_management_service/handlers/user_test.clj:
```diff
@@ -80,12 +81,44 @@
-             (:flash response))))))
+             (:flash response)))))
+  (with-redefs [client/post (constantly {:body (json/json-str {"success" true})})]
+    (testing "Should create a user who has a invite and associate the user to org of host"
```
See lines 69-78

yogirajh007 committed:
Fix date-generator to avoid testing errors

yogirajh007 committed:
Remove bug- change usages-done to usage-limit in invite-view

yogirajh007 committed:
Add support for transactions in core-db

- Transactions in user db & model

yogirajh007 committed:
Add checks for usage-limit and invite-validity

- Use transactions for enforcing usage-limits

shhivam commented on test/winter_onboarding_2021/fleet_management_service/db/user_test.clj:
```diff
@@ -79,3 +79,20 @@
+
+(deftest user-with-invitation
+  (testing "Should create an user with an invitation")
+  (let [admin (factories/admin)
+        _ (org-models/create-and-associate db-core/db-conn {:organisations/name "foo-org"}
+                                           admin)
+        invite (factories/invite-admin {:invites/created-by (:users/id admin)})
+        host (first (user-db/find-by-keys {:users/email (:users/email admin)}))
+        user (factories/user {:users/role (:invites/role invite)
+                                  :users/invite-id (:invites/id invite)
+                                  :users/org-id (:users/org-id host)})
+        _ (user-db/create user)
+        new-user (user-db/find-by-keys {:users/email (:users/email user)})]
```
`:users/id` `:users/email` can be destructured on line 85 itself

alishamohanty commented on src/winter_onboarding_2021/fleet_management_service/views/layout.clj:
```diff
@@ -8,6 +8,10 @@
    (include-js "/public/js/bootstrap.bundle.min.js")
+   [:script
+    {:defer "defer"
+     :async "async"
```
Shall we consider not including the script in layout, since it will be loaded on every request whereas we only require in invites only.

alishamohanty replied:
Please go through [this](https://developers.google.com/recaptcha/docs/display#:~:text=The%20script%20must%20be%20loaded%20using%20the%20HTTPS%20protocol%20and%20can%20be%20included%20from%20any%20point%20on%20the%20page%20without%20restriction). We can add the script in the signup form itself.

shhivam commented on test/winter_onboarding_2021/fleet_management_service/handlers/user_test.clj:
```diff
@@ -10,63 +14,164 @@
 (deftest user-login
   (testing "Correct login credentials, should redirect to user dashboard"
-    (with-redefs [utils/uuid (fn [] (java.util.UUID/fromString "9088992d-d0f4-4207-9b95-c934ad071c32"))]
+    (with-redefs [utils/uuid (fn [] (java.util.UUID/fromString "9088992d-d0f4-4207-9b95-c934ad071c32"))
+                  client/post (constantly {:body (json/json-str {"success" true})})]
       (let [user {:name "Severus Snape"
                   :email "s.snape@hogwarts.edu"
                   :role "admin"
-                  :password "lily"}
+                  :password "lily"
+                  :g-recaptcha-response "abcd123"}
             _ (handler/create-user {:form-params user})
```
In this `with-redefs`, it returns a success message regardless of the `:g-captcha-response` in the `user` hashmap. Could we check that `:g-captcha-response` is actually `"abcd123"` inside that redefinition?

Why? Because I couldnt figure out where `client/post` is being called in the code & then it dawned on me that we are doing it inside the handler. Doing the above might help in readability?

shhivam replied:
Since we are doing this redefinition in other tests as well, could we just stub this `client/post` function and use it everywhere redefinition is needed?

yogirajh007 replied:
Can you elaborate more?

shhivam commented on test/winter_onboarding_2021/fleet_management_service/models/user_test.clj:
```diff
@@ -24,6 +24,23 @@
 
+(deftest user-with-invitation
+  (testing "Should create an user with an invitation with org-id set to that of host")
+  (let [db-admin (factories/admin)
+        _ (org-models/create-and-associate db-core/db-conn {:organisations/name "foo-org"}
+                                           db-admin)
+        host (first (user-model/find-by-keys {:users/email (:users/email db-admin)}))
+        invite (factories/invite-manager {:invites/created-by (:users/id db-admin)})
+        new-user (factories/user {:users/role (:invites/role invite)
```
Same regarding the destructuring re: `db-admin`

shhivam commented on test/winter_onboarding_2021/fleet_management_service/models/user_test.clj:
```diff
@@ -24,6 +24,23 @@
+                                           db-admin)
+        host (first (user-model/find-by-keys {:users/email (:users/email db-admin)}))
+        invite (factories/invite-manager {:invites/created-by (:users/id db-admin)})
+        new-user (factories/user {:users/role (:invites/role invite)
+                                  :users/invite-id (:invites/id invite)
+                                  :users/org-id (:users/org-id host)})
+        _ (user-model/create new-user)
```
Could we use `factories/create` here?

shhivam commented on src/winter_onboarding_2021/fleet_management_service/specs.clj:
```diff
@@ -22,7 +22,10 @@
 
 (defn- date-inst-gen []
-  (gen/return (eval (sqltime/to-sql-time (cljt/from-now (cljt/weeks (rand-int 10)))))))
+  (gen/return (eval (sqltime/to-sql-time (cljt/from-now (cljt/weeks (+ 1 (rand-int 10))))))))
+
+(defn- capital-string-gen []
+  (gen/return (apply str (take 8 (repeatedly #(char (+ (rand 26) 65)))))))
 
```
Can we use threading macro here?

shhivam commented on src/winter_onboarding_2021/fleet_management_service/specs.clj:
```diff
@@ -74,19 +77,31 @@
+(s/def :users/invite-id uuid?)
+(s/def :users/org-id uuid?)
 
 (s/def ::users
   (s/keys :req [:users/name
                 :users/role
                 :users/email
                 :users/password]))
 
+(s/def ::user-with-invite-token
+  (s/keys :req [:users/name
+                :users/role
+                :users/email
+                :users/password
+                :users/org-id
+                :users/invite-id]))
+
 (s/def ::users-all-attr
   (s/keys :req [:users/id
                 :users/name
                 :users/role
                 :users/email
-                :users/password]))
+                :users/password
+                :users/invite-id
+                :users/org-id]))
 
```
The code in these specs are getting a little repetitive. Could use `clojure.spec.alpha/merge`? 

https://clojuredocs.org/clojure.spec.alpha/merge

yogirajh007 replied:
Tried using this, was breaking some changes, will look at this in a saperate PR later.

shhivam commented on src/winter_onboarding_2021/fleet_management_service/handlers/user.clj:
```diff
@@ -1,38 +1,98 @@
+(defn- user-exist? [email]
   (not-empty (user-model/find-by-keys {:users/email email})))
 
+(defn- recaptcha-invalid? [g-recaptcha-response]
+  (try (let [res (client/post "https://www.google.com/recaptcha/api/siteverify"
```
Should we add this hardcoeded URL in some config or constants file?

yogirajh007 replied:
Used only once, but still we can consider.

yogirajh007 commented on .github/workflows/terraform.yml:
```diff
@@ -18,6 +18,7 @@ jobs:
       AWS_SECRET_ACCESS_KEY:  ${{ secrets.AWS_SECRET_ACCESS_KEY }}
       TF_VAR_heroku_email: ${{ secrets.HEROKU_EMAIL }}
       TF_VAR_heroku_api_key: ${{ secrets.HEROKU_API_KEY }}
+      RECAPTCHA: ${{ secrets.RECAPTCHA }}
```
We will need it on the heroku server as well.

shhivam replied:
> We will need it on the heroku server as well.

That's true but this file doesnt have anything related to captcha, right?
Terraform needs heroku creds because it needs to issue new servers or destroy some other resources.

yogirajh007 replied:
Okay, but how will heroku get the env variables then?

shhivam replied:
It should not affect heroku in any way if we remove this line 21

yogirajh007 committed:
Remove erroneous RECAPTCHA env from terrafrom.yml

yogirajh007 committed:
Use config for recaptcha url & site-key

yogirajh007 committed:
Move recaptcha script from layout to user-view

yogirajh007 committed:
Implement suggestions on PR
- Use thread operator in specs/capital-string-gen
- Change formating of multi-arity function in user-db/find-by-keys
- Use destructuring in db/user-test for better readability
- Change assertion in views/user-test for better readability
